sugondat-chain/runtimes/sugondat-kusama/src/fee_adjustment_tests.rs

```diff
@@ -0,0 +1,617 @@
+        );
+
+        let (normal, max, target) =
+            if normalized_weight_dimensions.0 < normalized_weight_dimensions.1 {
+                (
+                    block_weight.proof_size(),
+                    max.proof_size(),
+                    target_weight.proof_size(),
+                )
+            } else {
+                (
+                    block_weight.ref_time(),
+                    max.ref_time(),
+                    target_weight.ref_time(),
+                )
+            };
+
+        // maximum tx weight
+        let m = max as f64;
+        // block weight always truncated to max weight
+        let block_weight = (normal as f64).min(m);
+        let v: f64 = self.adjustment_variable().to_float();
+
+        // ideal saturation in terms of weight
+        let ss = target as f64;
+        // current saturation in terms of weight
+        let s = block_weight;
+
+        let t1 = v * (s / m - ss / m);
+        let t2 = v.powi(2) * (s / m - ss / m).powi(2) / 2.0;
+        let next_float = previous_float * (1.0 + t1 + t2);
+        Multiplier::from_float(next_float)
+    }
+
+    fn error(&self) -> Multiplier {
+        match self {
+            // TODO: the error needs to be so high because probably
+            // something is broken in the implementation of c_traffic equation
+            MultiplierType::Length => Multiplier::from_inner(100000000000000000),
+            MultiplierType::Fee => Multiplier::from_inner(100),
+        }
+    }
+
+    fn adjustment_variable(&self) -> Multiplier {
+        match self {
+            MultiplierType::Length => AdjustmentVariableBlockSize::get(),
+            MultiplierType::Fee => AdjustmentVariableBlockFullness::get(),
+        }
+    }
+
+    fn target_percentage(&self) -> Multiplier {
+        match self {
+            MultiplierType::Length => TargetBlockSize::get().into(),
+            MultiplierType::Fee => TargetBlockFullness::get().into(),
+        }
+    }
+
+    // k = 1 / (v * s*)
+    fn k_bound(&self) -> u64 {
+        (Multiplier::saturating_from_integer(1)
+            / (self.target_percentage() * self.adjustment_variable()))
+        .to_float()
+        .ceil() as u64
+    }
+
+    fn run_with<F>(&self, w: Weight, assertions: F)
+    where
+        F: Fn() -> (),
+    {
+        let mut t: sp_io::TestExternalities = frame_system::GenesisConfig::<Runtime>::default()
+            .build_storage()
+            .unwrap()
+            .into();
+        t.execute_with(|| {
+            match self {
+                MultiplierType::Length => {
+                    // AllExtrinsicsLen is a private storage value of the system pallet
+                    // so the key must be manually constructed
+                    let len = w.ref_time();
+                    sp_io::storage::set(
+                        &[
+                            sp_core::twox_128(b"System"),
+                            sp_core::twox_128(b"AllExtrinsicsLen"),
```

Comment:
This probably should be upstreamed.

Reply:
What do you mean? Are you saying that AllExtrinsicsLen should not be private or that there should be a public helper to set the value of that StorageValue?

Reply:
I believe @pepyakin means an upstream PR to Polkadot-SDK

Reply:
Looking into `polkadot-sdk/master/src/frame_system/lib.rs` to understand how many things would be required to make the PR, I've noticed this:

```rust
	pub fn set_block_consumed_resources(weight: Weight, len: usize) {
		BlockWeight::<T>::mutate(|current_weight| {
			current_weight.set(weight, DispatchClass::Normal)
		});
		AllExtrinsicsLen::<T>::put(len as u32); // <-----
	}
```

This means that I don't need to manually construct the key, and I can simply use this method!